### PR TITLE
fault tolerant parse implementation

### PR DESCRIPTION
**Pull request by oluiscabral**

### CODE REVIEW

File: Cython/Compiler/TreeFragment.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ def find_module(self, module_name, relative_to=None, pos=None, need_pxd=1, absol
         return ModuleScope(module_name, parent_module=None, context=self)
 
 
-def parse_from_strings(name, code, pxds=None, level=None, initial_pos=None,
+def parse_from_strings(name, code, fault_tolerant=False, pxds=None, level=None, initial_pos=None,
```
Comment (fabioz) [inline Cython/Compiler/TreeFragment.py]: Add `fault_tolerant=False` as the last parameter (otherwise it'd change the api and break callers).

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 2]
```diff
@@ -59,6 +60,26 @@ def test_temps(self):
...
+def method3():
+    error_here =
+
+def method4():
+    without_error_here = 10
+'''
+        with local_errors() as errors:
+            parse_from_strings("test_name", code, True)
+
+        self.assertEqual("CompileError((<StringSourceDescriptor:test_name>, 6, 16), 'Expected an identifier or literal')", repr(errors[0]))
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py:99]: Please also check that `len(errors) == 3` and that `error[1]` is also what you expect.

File: Cython/Compiler/Parsing.py [hunk 3]
```diff
@@ -2337,14 +2338,25 @@ def p_statement(s, ctx, first_statement = 0):
 
 def p_statement_list(s, ctx, first_statement = 0):
     # Parse a series of statements separated by newlines.
+    global stat
```
Comment (fabioz) [inline Cython/Compiler/Parsing.py]: Why is this added?

File: Cython/Compiler/Parsing.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 
 # This should be done automatically
 import cython
+from Cython.Compiler.Errors import report_error, CompileError
```
Comment (fabioz) [inline Cython/Compiler/Parsing.py:10]: Please add a blank line after the import (as it was before).

File: Cython/Compiler/Parsing.py [hunk 5]
```diff
@@ -2337,14 +2338,25 @@ def p_statement(s, ctx, first_statement = 0):
...
+    global stat
     pos = s.position()
     stats = []
+
     while s.sy not in ('DEDENT', 'EOF'):
-        stat = p_statement(s, ctx, first_statement = first_statement)
+        try:
+            stat = p_statement(s, ctx, first_statement = first_statement)
+        except Exception as exc:
+            if s.fault_tolerant == True:
```
Comment (fabioz) [inline Cython/Compiler/Parsing.py]: You don't need to check if `fault_tolerant == True`, just do `if s.fault_tolerant:`.

File: Cython/Compiler/Scanning.py [hunk 6]
```diff
@@ -301,7 +301,7 @@ class PyrexScanner(Scanner):
     #  compile_time_eval  boolean  In a true conditional compilation context
     #  compile_time_expr  boolean  In a compile-time expression context
 
-    def __init__(self, file, filename, parent_scanner=None,
+    def __init__(self, file, filename, fault_tolerant, parent_scanner=None,
```
Comment (fabioz) [inline Cython/Compiler/Scanning.py]: Add `fault_tolerant=False` as the last parameter (otherwise it'd change the api and break callers).

File: Cython/Compiler/Parsing.py [hunk 7]
```diff
@@ -2337,14 +2338,25 @@ def p_statement(s, ctx, first_statement = 0):
...
+                report_error(CompileError(pos, "Internal Error: "+ str(exc)))
```
Comment (fabioz) [inline Cython/Compiler/Parsing.py]: Please add a comment elaborating a bit on why this is needed...

Something as:

```
            # When processing statements in a fault-tolerant mode, it's possible
            # that because some Node isn't properly constructed, internal errors
            # happen (for instance if the right hand side of an assignment would
            # be `None`, trying to access its `pos` would throw an
            # AttributeError). For the fault-tolerant mode this is ok (we don't
            # want to change the whole parser to accommodate that use-case, so,
            # just report that as a `CompileError` so that it can be seen and
            # proceed as usual).
```

Comment (fabioz) [inline Cython/Compiler/Parsing.py]: We should also check if it's not already a `CompileError`, just in case.

i.e.:
```
if not isinstance(exc, CompileError):
    exc = CompileError(pos, "Internal Error: "+ str(exc)
report_error(exc)
```

File: Cython/Compiler/Parsing.py [hunk 8]
```diff
@@ -2337,14 +2338,25 @@ def p_statement(s, ctx, first_statement = 0):
...
+                s.next()
+            else:
+                raise exc
+
         if isinstance(stat, Nodes.PassStatNode):
             continue
         stats.append(stat)
         first_statement = False
+
```
Comment (fabioz) [inline Cython/Compiler/Parsing.py]: Remove extraneous blank line added (we should try to touch as little code as possible for a pull request in a different project as that's probably the project standard).

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 9]
```diff
@@ -59,6 +60,26 @@ def test_temps(self):
...
+    pass
+
+def method2():
+    error_here =
+
+def method3():
+    error_here =
+
+def method4():
+    without_error_here = 10
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py]: I think that the last statement should also be an error (say `error_here =`) to see how it behaves in that corner case.

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 10]
```diff
@@ -59,6 +60,26 @@ def test_temps(self):
...
+    def test_parse_fault_tolerant_complete(self):
+        code = u'''
+def method():
+    pass
+
+def method2():
+    error_here =
+
+def method3():
+    error_here =
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py]: Maybe we could do some different error here (say just `error_here.`)

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 11]
```diff
@@ -59,6 +60,26 @@ def test_temps(self):
...
+
+def method4():
+    without_error_here = 10
+'''
+        with local_errors() as errors:
+            parse_from_strings("test_name", code, True)
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py]: Please verify that the return is a valid node.

i.e.:

```
node = parse_from_strings("test_name", code, True)
# Verify that node is what's expected and that the methods are there.
```

**Review by fabioz — CHANGES_REQUESTED** [review]

File: Cython/Compiler/Tests/TestTreeFragment.py (hunk 11 above)
Comment (fabioz) [reply Cython/Compiler/Tests/TestTreeFragment.py]: i.e.: `method, method2, method3, method4`

File: Cython/Compiler/Parsing.py (hunk 3 above)
Comment (oluiscabral) [reply Cython/Compiler/Parsing.py]: because without it, stat raises an internal error of "reference before assignment" inside the try catch

-> error raised: "Internal Error: local variable 'stat' referenced before assignment"

Comment (fabioz) [reply Cython/Compiler/Parsing.py]: This isn't right. Please remove that.

I believe the proper fix is adding a 

`continue`

right after the `report_error`.

Comment (fabioz) [reply Cython/Compiler/Parsing.py]: p.s.: if the `continue` doesn't work we can discuss a bit more about it...

Comment (fabioz) [reply Cython/Compiler/Parsing.py]: Actually, it should be after the `report_error`, but after the `s.next()`.

Comment (oluiscabral) [reply Cython/Compiler/Parsing.py]: it worked fine ;)

File: Cython/Compiler/Parsing.py [hunk 12]
```diff
@@ -2340,7 +2342,25 @@ def p_statement_list(s, ctx, first_statement = 0):
...
+        # proceed as usual).
+        stat = None
+        try:
+            stat = p_statement(s, ctx, first_statement = first_statement)
+        except Exception as exc:
+            if s.fault_tolerant:
+                if not isinstance(exc, CompileError):
+                    exc = CompileError(pos, "Internal Error: "+ str(exc))
+                report_error(exc)
+                s.next()
```
Comment (fabioz) [inline Cython/Compiler/Parsing.py:2360]: Put a `continue` here so that the node is not added (that way you also don't need the `stat=None` assignment in the start of this function).

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 13]
```diff
@@ -2,7 +2,9 @@
 from Cython.Compiler.TreeFragment import *
 from Cython.Compiler.Nodes import *
 from Cython.Compiler.UtilNodes import *
+from Cython.Compiler.Errors import local_errors
 import Cython.Compiler.Naming as Naming
+from nt import stat
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py]: This shouldn't be here.

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 14]
```diff
@@ -59,6 +61,39 @@ def test_temps(self):
...
+        self.assertIsInstance(stats[0], DefNode)
+        self.assertIsInstance(stats[0].body, PassStatNode)
+        self.assertIsInstance(stats[1], DefNode)
+        self.assertIsNone(stats[1].body)
+        self.assertIsInstance(stats[2], DefNode)
+        self.assertIsInstance(stats[2].body, ExprStatNode)
+        self.assertIsInstance(stats[3], DefNode)
+        self.assertIsInstance(stats[3].body, SingleAssignmentNode)
+        self.assertIsInstance(stats[4], DefNode)
+        self.assertIsNone(stats[4].body)
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py]: There should be no `None` entries in lists (the `Node` should be skipped in this case).

Comment (fabioz) [reply Cython/Compiler/Tests/TestTreeFragment.py]: Note: the fix to use the `continue` should fix this...

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 15]
```diff
@@ -59,6 +61,39 @@ def test_temps(self):
...
+            stats = parse_from_strings("test_name", code, fault_tolerant=True).body.stats
+        self.assertIsInstance(stats[0], DefNode)
+        self.assertIsInstance(stats[0].body, PassStatNode)
+        self.assertIsInstance(stats[1], DefNode)
+        self.assertIsNone(stats[1].body)
+        self.assertIsInstance(stats[2], DefNode)
+        self.assertIsInstance(stats[2].body, ExprStatNode)
+        self.assertIsInstance(stats[3], DefNode)
+        self.assertIsInstance(stats[3].body, SingleAssignmentNode)
+        self.assertIsInstance(stats[4], DefNode)
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py:95]: It'd be nice that you can check the node name too to know which method was recognized.

File: Cython/Compiler/Tests/TestTreeFragment.py (hunk 2 above)
Comment (fabioz) [reply Cython/Compiler/Tests/TestTreeFragment.py:99]: This was missed in the updated version.

File: Cython/Compiler/Tests/TestTreeFragment.py (hunk 15 above)
Comment (fabioz) [reply Cython/Compiler/Tests/TestTreeFragment.py:95]: Also, please check `len(stats)` to know how many statements were generated.

File: Cython/Compiler/Tests/TestTreeFragment.py (hunk 2 above)
Comment (fabioz) [reply Cython/Compiler/Tests/TestTreeFragment.py:99]: (the `len()` part)

File: Cython/Compiler/Parsing.py (hunk 12 above)
Comment (oluiscabral) [reply Cython/Compiler/Parsing.py:2360]: I notice that using `continue` without assigning stat a value causes errors ahead on the code. If it continue the loop without changing stat var, it will store wrong values at the AST

example: if the statement before the errored statement is a PassStatementNode, the errored statement will also be stored as a PassStatementNode

Comment (oluiscabral) [reply Cython/Compiler/Parsing.py:2360]: The code ahead needs the stat value, even if it is None, because the DefStatement is caught and it need to have a body type. What is raising errors to get identified is the statement inside the def, not the DefStatement itself

File: Cython/Compiler/Tests/TestTreeFragment.py (hunk 2 above)
Comment (oluiscabral) [reply Cython/Compiler/Tests/TestTreeFragment.py:99]: Alright

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 16]
```diff
@@ -59,6 +60,46 @@ def test_temps(self):
...
+
+def method4():
+    without_error = 10
+
+def method5():
+    error_here =
+'''
+        with local_errors() as errors:
+            stats = parse_from_strings("test_name", code, fault_tolerant=True).body.stats
+        self.assertTrue(len(stats) == 5)
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py]: It should be `self.assertEqual(len(stats), 5)`

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 17]
```diff
@@ -59,6 +60,46 @@ def test_temps(self):
...
+        self.assertIsInstance(stats[2], DefNode)
+        self.assertEqual(str(stats[2].name), "method3")
+        self.assertIsInstance(stats[2].body, ExprStatNode)
+        self.assertIsInstance(stats[3], DefNode)
+        self.assertEqual(str(stats[3].name), "method4")
+        self.assertIsInstance(stats[3].body, SingleAssignmentNode)
+        self.assertIsInstance(stats[4], DefNode)
+        self.assertEqual(str(stats[4].name), "method5")
+        self.assertIsInstance(stats[4].body, PassStatNode)
+        self.assertTrue(len(errors) == 5)
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py]: It should be `self.assertEquals(len(errors), 5)`

File: Cython/Compiler/Tests/TestTreeFragment.py [hunk 18]
```diff
@@ -59,6 +60,46 @@ def test_temps(self):
...
+        self.assertEqual("CompileError((<StringSourceDescriptor:test_name>, 6, 16), 'Expected an identifier or literal')", repr(errors[0]))
+        self.assertEqual("CompileError((<StringSourceDescriptor:test_name>, 9, 15), 'Expected an identifier')", repr(errors[2]))
+        self.assertEqual("CompileError((<StringSourceDescriptor:test_name>, 15, 16), 'Expected an identifier or literal')", repr(errors[3]))
```
Comment (fabioz) [inline Cython/Compiler/Tests/TestTreeFragment.py:102]: Please check all errors (including internal errors at pos 1 and 4).

File: Cython/Compiler/Parsing.py [hunk 19]
```diff
@@ -2340,7 +2342,25 @@ def p_statement_list(s, ctx, first_statement = 0):
     pos = s.position()
     stats = []
     while s.sy not in ('DEDENT', 'EOF'):
-        stat = p_statement(s, ctx, first_statement = first_statement)
+        # When processing statements in a fault-tolerant mode, it's possible
```
Comment (fabioz) [inline Cython/Compiler/Parsing.py]: Please put this comment right after the `if s.fault_tolerant`.

Comment (oluiscabral) [reply Cython/Compiler/Parsing.py]: with scope identation?
